Specify tool and configuration notifications

lgolding opened the pull request:
Also:
- Remove `result.shortMessage` and rename `result.fullMessage` to `message`.
- Rename `result.formattedMessage` to `formattedRuleMessage`.

@michaelcfanning

Commit:
Specify tool and configuration notifications

Also:
* Remove `result.shortMessage` and rename `result.fullMessage` to `message`.
* Rename `result.formattedMessage` to `formattedRuleMessage`.

lgolding commented on document/AnnexC.mdk:
```diff
@@ -50,6 +50,8 @@ allows viewer implementers to assume that all files from the same version of the
 
 This general guidance is embodied in various sections of the specification. For example:
 
+* A converter should not attempt to synthesize a `ruleId` for a result if the tool does not provide one.
```
I found another example in the of the guidance "don't synthesize stuff", and replicated it here in the Annex about converters.

lgolding commented on document/CoreFormat.mdk:
```diff
@@ -2225,3 +2260,112 @@ of the resulting UTF-8 byte sequence.
+A `notification` object may contain a property named `id` whose value is a string containing
+the stable, opaque identifier for the condition that was encountered.
+
+### `ruleId` property { #notification-ruleId }
```
I propose to add this.

michaelcfanning replied:
I see. So is a notification happens to be raised in some context that relates directly to a rule, we provide a placeholder for that data. Yes. Solves a known issue we've encountered in production (rule 'foo' crashed, for example.). OK, agree

lgolding commented on document/CoreFormat.mdk:
```diff
@@ -2225,3 +2260,112 @@ of the resulting UTF-8 byte sequence.
+
+If the `level` property is absent, it shall be considered equivalent to the value `warning`.
+
+### `time` property { #notification-time }
```
I propose to add this.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -413,6 +413,44 @@ The `results` array shall be empty if the run that produced the `run` object did
+If the error occurs in the course of evaluating a rule, the tool might report the error in `toolNotifications`, disable the rule,
+and continue to execute the remaining rules.
+
+If the error occurs outside of the evaluation of a rule, the tool might report
```
i don't understand what you're saying here. the tool might report an error and halt. or it might not report an error and then exit? don't we want tools to report when they halt unexpectedly?

lgolding replied:
This was existing text. What I wanted to convey was that the tool might just plain crash before it has time to report the error. Of course, I said "either normally or abnormally", which sounds like it's ok for tools not to report the problem even if they _do_ have the chance to report.

How about replacing the last sentence with "If the tool exits abnormally, it might not have the opportunity to report the notification."?

lgolding replied:
Fixed.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -413,6 +413,44 @@ The `results` array shall be empty if the run that produced the `run` object did
+### `toolNotifications` { #run-toolNotifications }
+
+A `run` object may contain a property named `toolNotifications`
+whose value is an array of zero or more `notification` objects (&sect;[#notification]).
```
There's nothing in here that addresses the other primary use for this section: to persist information for a tool developer to troubleshoot breaks and crashes. Maybe this comes later?

lgolding replied:
No, you're right, I completely failed to point out the whole "audience" thing. Will fix.

lgolding replied:
Fixed.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -413,6 +413,44 @@ The `results` array shall be empty if the run that produced the `run` object did
+and continue to run.
+~
+
+~ IsoExample
```
Again, can we sneak in the point that this information is targeted towards the person or system responsible for producing tool runs? not the tool owner. not necessarily the tool results consumers (but could be, of course, when a user is running a tool for her own purposes)

lgolding replied:
Again, you're right. I'm not going to "sneak it in" -- it's kind of the whole point, and I completely failed to say it.

lgolding replied:
Fixed.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -413,6 +413,44 @@ The `results` array shall be empty if the run that produced the `run` object did
+~ IsoExample
+Suppose a tool is invoked with a configuration file which specifies that the tool should disable rule X.
+If there is no such rule as rule X, then the tool should report it in `configurationNotifications`. The tool might
+continue to run, reporting results for the rules that are correctly configured.
```
What about a bad command-line argument? An output file path where a file already exists or that is inaccessible? I feel like we could use more examples (i.e., ones that are errors and block execution). I like your example for a configuration warning, where analysis clearly could continue (and probably should)

lgolding replied:
Sure, I can add more examples.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -775,26 +813,27 @@ A `result` object describes a single result produced by an analysis tool.
 
-A `result` object should contain a property named `ruleId` whose value is a string containing
+If the `level` property (&sect;[#result-level]) of a `result` object is anything other than `note`,
+then the `result` object shall contain a property named `ruleId` whose value is a string containing
```
Wait, we say the id is required 'shall' but then note a converter might not be able to provide one. ??

lgolding replied:
I'll change it to say that if the log is produced by an analysis tool as opposed to a converter, it shall be present.

lgolding replied:
Fixed

---
In reply to: [61356581](https://github.com/sarif-standard/sarif-spec/pull/172#discussion_r61356581) [](ancestors = 61356581)

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -775,26 +813,27 @@ A `result` object describes a single result produced by an analysis tool.
-Analysis tools that produce the SARIF format shall provide a `ruleId` for each `result`
-(with the possible exception of `result`s whose `kind` property is `note`; see (&sect;[#result-kind]).
+If the `level` property of the `result` object is `note`,
+then the `ruleId` property may be absent.
```
Is there a typo a couple of lines below this: 
shall omit 'ruleId' from the 'result's.

extra 's' at the end of that sentence?

lgolding replied:
Not a typo. It renders `result` in the code font and then adds the "s":

`result`s

lgolding replied:
Not pretty here but looks nicer in the HTML.

lgolding replied:
In any case, moot, since I rephrased the sentence and it no longer ends that way.

---
In reply to: [61357523](https://github.com/sarif-standard/sarif-spec/pull/172#discussion_r61357523) [](ancestors = 61357523)

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -805,83 +844,92 @@ If present, the `kind` property shall have one of the following values, with the
+        {
+            "ruleId": "ABC0001",
+            "level": "notApplicable",
+            "message": "MyTool64.exe is a 64-bit binary, so this rule does not apply to it."
```
Can we provide a better message here rather than one that we would file a bug against if it shipped in a real tool? :)

'MyTool64.exe' was not evaluated for check ABC1001 as it is not a 32-bit binary.

or some such

lgolding replied:
ok

lgolding replied:
Fixed

---
In reply to: [61357436](https://github.com/sarif-standard/sarif-spec/pull/172#discussion_r61357436) [](ancestors = 61357436)

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -2225,3 +2260,112 @@ of the resulting UTF-8 byte sequence.
+
+### `id` property { #notification-id }
+
+A `notification` object may contain a property named `id` whose value is a string containing
```
I don't think we have precisely the same requirement for a stable opaque id here. I think the id is anything the tool owner feels is appropriate to categorize/locate relevant messages. The stability of rule ids is encouraged because these get burned into baselines. Because the ids must be stable, we encourage an opaque id for rules, otherwise you might need to stick with a readable description that becomes inaccurate over time as the rule implementation evolves. 

For notification ids, do what you will noting that its worth giving some thought into how/whether automation tools might want to locate specific problems.

lgolding replied:
I'd like to leave this as is. The requirement for tools to parse logs makes it useful for the id to be stable. And "opaque" doesn't mean "unreadable by a human", it means "unparsable by a machine" (that is, a machine should not assume that the identifier has any particular internal structure or syntax).

---
In reply to: [61357910](https://github.com/sarif-standard/sarif-spec/pull/172#discussion_r61357910) [](ancestors = 61357910)

michaelcfanning replied:
No, we should change this, the requirements around the readable and non-readable ids are very clear based on background with PREfast and FxCop and they don't apply to tool errors. In that context, 'opaque' here certainly does mean that the id doesn't communicate information to a human. A rule requires a stable identifier for a couple of clear reasons: 1) so that specific checks can be enabled/disabled in scripts, etc., 2) to refer to in baselines to track ongoing review. If you change a rule id, you risk breaking lab infrastructure or reraising issues to users that they have seen before. Because an id needs to be extremely stable, it shouldn't try to convey any meaningful information to users that could change over time. A rule 'DoNotFooOrBar' might have a change in implementation at some point, because a security issue in 'Bar' was resolved. Meanwhile, the 'DoNotFooOrBar' guidance can persist in baselines, in-source suppressions, etc., forever conveying outdated or inaccurate guidance. 

A tool error is a different beast. Tools should not provide the ability to selectively disable them or to ignore tool errors in baselines: any notification with a severity of error should fail, period. Tool developers are free to change these ids at any time or to convey whatever information makes sense. Think of all this like the result of GetHashCode implementation. Users should not take a dependency on a specific GetHashCode result forever expressing something about the associated object. That's a bug. That id is specific/internal to the class. Same thing with a tool error id.

lgolding replied:
Fixed.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -2225,3 +2260,112 @@ of the resulting UTF-8 byte sequence.
+
+If present, the `level` property shall have one of the following values, with the specified meanings:
+
+* `error`: A serious problem was found.
```
Do we want to provide more guidance? An error represents a condition where analysis halted or where results might be fundamentally compromised. a warning is appropriate when there's a lack of certainty around whether an issue occurred or where analysis may be incomplete but the results that were generated are probably valid. Warnings represent conditions that are either assumed to be transient or which really should be looked at (i.e., they are things which can and should be cleaned up in a perfect world of adequate resources). 'info' messages imply no requirement for action.

lgolding replied:
I adopted most of this, except the language around "transient" (errors can be transient too).

---
In reply to: [61358382](https://github.com/sarif-standard/sarif-spec/pull/172#discussion_r61358382) [](ancestors = 61358382)

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -2225,3 +2260,112 @@ of the resulting UTF-8 byte sequence.
+
+### `time` property { #notification-time }
+
+A `notification` object may contain a property named `time` whose value is a string
```
ok, like 'time'. please note that when we author the section of providing run-over-run ordering, we should recommend that all notifications be sorted by timestamp, if present.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -2225,3 +2260,112 @@ of the resulting UTF-8 byte sequence.
+
+### `exception` property { #notification-exception }
+
+If the issue is a result of a runtime exception, the `notification` object may contain a property named `exception`
```
You are sprinkling the term 'issue' around quite consciously, yes?

lgolding replied:
No, it's a holdover from an earlier naming round when we were calling these `toolIssues` and `configurationIssues`. Will change to "notification".

lgolding replied:
Fixed.

---
In reply to: [61360057](https://github.com/sarif-standard/sarif-spec/pull/172#discussion_r61360057) [](ancestors = 61360057)

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -2225,3 +2260,112 @@ of the resulting UTF-8 byte sequence.
+each of which is considered to be a cause of the containing `exception`.
+
+~ IsoNote
+There is commonly no more than one inner exception.
```
Rather than suggest this is a special-case to accommodate .NET, we should note that this exists to accommodate platforms that provide a mechanism for aggregating exceptions, such as the .NET AggregateException type. This is useful functionality

lgolding replied:
Fixed.

---
In reply to: [61359372](https://github.com/sarif-standard/sarif-spec/pull/172#discussion_r61359372) [](ancestors = 61359372)

Commit:
Partial #169: Address some of the PR feedback

Commit:
Address remaining PR feedback on #172

Commit:
Rename file.offsetFromParent to file.offset

Should have done this at the same time we changed file.pathFromParent to
file.uri. The schema correctly has both file.uri and file.offset.

Commit:
Revise guidance for notification.id

... and add explanation for why rule.id is stable and opaque,
but notification.id need not be (per @michaelcfanning).

lgolding commented:
@michaelcfanning, please take a look at the existing description of `rule.name`, and the NOTE that accompanies it, and tell me if that covers it.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -413,6 +413,99 @@ The `results` array shall be empty if the run that produced the `run` object did
+that the run failed.
+
+~ IsoNumberedNote { @IsoNumberedNote=1 }
+The information in `toolNotifications` is primarily intended for the developers of the analysis tool,
```
Yes, this is the primary audience, with one additional note that viewers/automation systems will minimally want to consult these sections to ensure they don't contain errors (and to communicate that fact to users if they do). Makes sense? the content/detail in the sections are targeted towards the tool owner, but the results consumer, minimally, needs to be alerted that this content exists (particularly for errors).

lgolding replied:
Fixed.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -413,6 +413,99 @@ The `results` array shall be empty if the run that produced the `run` object did
+that the run failed.
+
+~ IsoNumberedNote { @IsoNumberedNote=1 }
+The information in `toolNotifications` is primarily intended for the engineers who configure the analysis tool,
```
s/be configurationNotifications

lgolding replied:
Fixed.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -413,6 +413,99 @@ The `results` array shall be empty if the run that produced the `run` object did
+        "id": "UnknownRule",
+        "ruleId": "ABC0001",
+        "level": "warning",
+        "message": "The configuration file specifies that an unknown rule \"ABC0001\" is to be disabled." 
```
In practice as a tools developer, I would simpler, active voice language for this.
'Could not located specified rule 'ABC0001' in order to disable it.'

lgolding replied:
How about "Could not disable rule ABC0001 because there is no rule with that id"

michaelcfanning replied:
Even better. Put 'ABC2001' in quotes, our common convention for the dynamic content within a formatted message.

lgolding replied:
Done.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -775,26 +868,31 @@ A `result` object describes a single result produced by an analysis tool.
 Not all existing analysis tools emit the equivalent of a `ruleId` in their output.
 A conversion tool which converts the output of such an analysis tool to the SARIF format
-shall omit `ruleId` from the `result`s.
+shall not set the `ruleId` property, and in particular, it shall not attempt to
```
Is it still clear, though, that if a tool (such as FxCop or cl.exe) provides an id that meets our definition, the converter should provide it? I think so but what do you think?...

lgolding replied:
Yes, I think it is clear, because the paragraph sets the context as "tools that do not emit the equivalent of a rule id".

lgolding replied:
I mean, the paragraph sets the context of the _exemption_ to only those tools.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -805,83 +903,91 @@ If present, the `kind` property shall have one of the following values, with the
+        {
+            "ruleId": "ABC0002",
+            "level": "note",
+            "message": "Consider using 'nameof(start)' instead of hard-coding the parameter name 'start'."
```
I continue to see this as a warning. This isn't purely information, you're telling the developer about a potential problem: they may rename parameter 'start' and neglected to update the hard-coded string literal referring to it. Using nameof will eliminate this potential maintainability problem. Good result. Not an 'info'. 'info' object provide observational information. this information may _support_ analysis or a result. but the item itself shouldn't constitute a discrete issue which you might want to track and fix. our experience says that three levels of priority for things you might fix is too much: the p3s are just noise. developers put things into two buckets: must fix, nice to fix. errs & warnings, iow.

lgolding replied:
Then we are missing the scenario where in VS, a user sets severity to "info". These are real observations -- real "results" -- but the user has decided it doesn't rise to the level of a warning. This is what I was saying yesterday when I wrote "note" and "info" on the board and asked if they weren't two distinct concepts.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1880,7 +1978,7 @@ A `rule` object contains information that describes a rule.
 Either the `shortDescription` property (&sect;[#rule-shortDescription])
 or the `fullDescription` property (&sect;[#rule-fullDescription])
-or both must be present.
+or both shall be present.
```
'both' shall be present? didn't we just whack the shortMessageu?

lgolding replied:
We whacked `result.shortMessage`. This is `rule.shortDescription` (which I still like).

michaelcfanning replied:
Oops. Yes. :)

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1891,6 +1989,23 @@ a stable, opaque identifier for the rule.
+~ IsoNote
+Rule identifiers must be stable for two reasons:
+
+1. So specific checks can be disabled in build automation scripts, without the risks
```
use 'referenced' instead of 'disabled' or say something like 'so specific checks can be referenced (e.g., to disable them)'

lgolding replied:
"So build automation scripts can refer to specific checks, for example, to disable them, without the risk of a script breaking if a rule id changes."

michaelcfanning replied:
Nice, yes.

michaelcfanning commented on document/CoreFormat.mdk:
```diff
@@ -1891,6 +1989,23 @@ a stable, opaque identifier for the rule.
+without erroneously designating results as "new" because a rule id has changed.
+
+Rule identifiers should be opaque -- that is, they should not convey information to
+a user -- because a rule's implementation might change over time.
```
Another common source of rule id instability has to do with refactoring. i.e. the collapse of checks into things that address a common weakness or the reverse, exploding a check that was too general.

lgolding replied:
As we did that recently with the FxCop XML rules. But I'm not sure how that fits into a note that explains why rule ids should be stable and opaque. In the scenario you're describing, the rule ids were _not_ in fact stable.

michaelcfanning replied:
Absolutely right. I agree it's difficult to see how this fits in, we would merely be making a point around the difficulties of defining/maintaining a stable id. Leave it out.

Commit:
Still more tweaks to "notifications"

Commit:
Remove exception.{properties,tags}
